### PR TITLE
CUtil: remove usage of ARRAY_SIZE macro

### PR DESCRIPTION
**Pull request by lrusak**
Replaces many raw arrays with std::array. This allows us to remove the ARRAY_SIZE macro.

Clang format had a go at some of the changes.
It's best viewed with whitespace changes disabled. You can view that here: [link](https://github.com/xbmc/xbmc/pull/20664/files?w=1)

### CODE REVIEW

**Comment by notspiff** [comment]
Yet clang format let the commit message slide?  :p

lør. 11. des. 2021, 19:47 skrev Lukas Rusak ***@***.***>:

File: xbmc/cores/AudioEngine/Sinks/AESinkPULSE.cpp [hunk 1]
```diff
@@ -145,31 +147,12 @@ static pa_encoding AEFormatToPulseEncoding(AEDataFormat format)
...
 
-static AEDataFormat defaultDataFormats[] = {
-  AE_FMT_U8,
-  AE_FMT_S16NE,
-  AE_FMT_S24NE3,
-  AE_FMT_S24NE4,
-  AE_FMT_S32NE,
-  AE_FMT_FLOAT
-};
+constexpr std::array<AEDataFormat, 6> defaultDataFormats = {
```
Comment (howie-f) [inline xbmc/cores/AudioEngine/Sinks/AESinkPULSE.cpp:150]: is it intended or should it be `static constexpr`, same below?

Comment (lrusak) [reply xbmc/cores/AudioEngine/Sinks/AESinkPULSE.cpp:153]: I've pushed another commit to add them to an anonymous namespace instead.

**Review by AlwinEsch — COMMENTED** [review]
Your change makes Kodi more correct to the C++ style and would be good to have.

Only few comments by me.

Think on separate requests can be this done on all places in Kodi.

File: xbmc/dbwrappers/mysqldataset.cpp [hunk 2]
```diff
@@ -717,28 +718,18 @@ struct StrAccum {
...
+    {'d', 10, 1, etRADIX, 0, 0},     {'s', 0, 4, etSTRING, 0, 0},
+    {'g', 0, 1, etGENERIC, 30, 0},   {'z', 0, 4, etDYNSTRING, 0, 0},
+    {'q', 0, 4, etSQLESCAPE, 0, 0},  {'Q', 0, 4, etSQLESCAPE2, 0, 0},
+    {'w', 0, 4, etSQLESCAPE3, 0, 0}, {'c', 0, 0, etCHARX, 0, 0},
+    {'o', 8, 0, etRADIX, 0, 2},      {'u', 10, 0, etRADIX, 0, 0},
+    {'x', 16, 0, etRADIX, 16, 1},    {'X', 16, 0, etRADIX, 0, 4},
+    {'f', 0, 1, etFLOAT, 0, 0},      {'e', 0, 1, etEXP, 30, 0},
+    {'E', 0, 1, etEXP, 14, 0},       {'G', 0, 1, etGENERIC, 14, 0},
+    {'i', 10, 1, etRADIX, 0, 0},     {'n', 0, 0, etSIZE, 0, 0},
+    {'%', 0, 0, etPERCENT, 0, 0},    {'p', 16, 0, etPOINTER, 0, 1},
```
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/dbwrappers/mysqldataset.cpp]: Yea I added it. We should probably figure out clang-format options to see if we can just get it to have all options on a newline.

File: xbmc/platform/android/storage/AndroidStorageProvider.cpp [hunk 3]
```diff
@@ -327,7 +328,7 @@ std::set<std::string> CAndroidStorageProvider::GetRemovableDrivesLinux()
 
           // What mount points are accepted
           bool mountok = false;
-          for (unsigned int i=0; i < ARRAY_SIZE(mountWL); ++i)
+          for (size_t i = 0; i < mountWL.size(); ++i)
```
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/platform/android/storage/AndroidStorageProvider.cpp]: Updated in a new commit, thanks!

File: xbmc/windowing/windows/WinKeyMap.h [hunk 4]
```diff
@@ -15,133 +15,135 @@
...
-      static void DIB_InitOSKeymap()
-      {
-        /* Map the VK keysyms */
-        for (int i = 0; i < ARRAY_SIZE(VK_keymap); ++i)
-          VK_keymap[i] = XBMCK_UNKNOWN;
+    static void DIB_InitOSKeymap()
+    {
+      /* Map the VK keysyms */
+      for (size_t i = 0; i < VK_keymap.size(); ++i)
+        VK_keymap[i] = XBMCK_UNKNOWN;
```
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/windowing/windows/WinKeyMap.h]: Thanks for the tip!

File: xbmc/windowing/windows/WinKeyMap.h [hunk 5]
```diff
@@ -172,7 +173,7 @@ namespace KODI
           VK_keymap[VK_LAUNCH_APP1] = XBMCK_LAUNCH_APP1;
           VK_keymap[VK_LAUNCH_APP2] = XBMCK_LAUNCH_APP2;
         }
-      }
+    }
```
Comment (fuzzard) [inline xbmc/windowing/windows/WinKeyMap.h:176]: This looks thrown off with the rest of the whitespace. May be best to clang-format everything past the ifdef VK_PRINT

Comment (lrusak) [reply xbmc/windowing/windows/WinKeyMap.h]: I decided to add a pre-commit that runs clang-format on the entire file. I think that makes the diff look a lot better.

**Comment by lrusak** [comment]
I'm not going to apply the clang-format diff for `WinEventsWin32.cpp` as it seem rearranging the includes breaks the build and I can't be bothered to figure windows stuff out.

File: xbmc/cores/VideoPlayer/DVDCodecs/DVDCodecUtils.cpp [hunk 6]
```diff
@@ -12,6 +12,8 @@
 #include "cores/FFmpeg.h"
 #include "cores/VideoPlayer/Interface/TimingConstants.h"
 
+#include <array>
```
Comment (fuzzard) [inline xbmc/cores/VideoPlayer/DVDCodecs/DVDCodecUtils.cpp:15]: I think Util.h can be removed in this file now.
Trying to look at the others, but its a bit of a mixed bag in that header, haha.

File: xbmc/platform/freebsd/network/NetworkFreebsd.cpp [hunk 7]
```diff
@@ -92,23 +91,18 @@ bool CNetworkInterfaceFreebsd::GetHostMacAddress(unsigned long host_ip, std::str
...
 
-  mib[0] = CTL_NET;
-  mib[1] = PF_ROUTE;
-  mib[2] = 0;
-  mib[3] = AF_INET;
-  mib[4] = NET_RT_FLAGS;
-  mib[5] = RTF_LLINFO;
-
-  if (sysctl(mib, ARRAY_SIZE(mib), NULL, &needed, NULL, 0) == 0)
+  if (sysctl(mib.data(), mib.size(), NULL, &needed, NULL, 0) == 0)
```
Comment (howie-f) [inline xbmc/platform/freebsd/network/NetworkFreebsd.cpp:100]: why not `const_cast<int*>(mib.data())` like in [NetworkOsx.cpp](https://github.com/xbmc/xbmc/blob/33fd69de20d89eabf98857048da031cfc510f879/xbmc/platform/darwin/osx/network/NetworkOsx.cpp#L72) ?

File: xbmc/platform/freebsd/network/NetworkFreebsd.cpp [hunk 8]
```diff
@@ -92,23 +91,18 @@ bool CNetworkInterfaceFreebsd::GetHostMacAddress(unsigned long host_ip, std::str
...
   {
     buf = (char*)malloc(needed);
     if (buf)
     {
-      if (sysctl(mib, ARRAY_SIZE(mib), buf, &needed, NULL, 0) == 0)
+      if (sysctl(mib.data(), mib.size(), buf, &needed, NULL, 0) == 0)
```
Comment (howie-f) [inline xbmc/platform/freebsd/network/NetworkFreebsd.cpp:105]: const_cast ?

File: xbmc/platform/freebsd/network/NetworkFreebsd.cpp (hunk 7 above)
Comment (lrusak) [reply xbmc/platform/freebsd/network/NetworkFreebsd.cpp:100]: I guess the freebsd implementation allows a `const` type.

**Review by howie-f — APPROVED** [review]
looks good to me, maybe just fuzzards header comment to address, nice

File: xbmc/cores/VideoPlayer/DVDCodecs/DVDCodecUtils.cpp (hunk 6 above)
Comment (lrusak) [reply xbmc/cores/VideoPlayer/DVDCodecs/DVDCodecUtils.cpp:14]: updated, thanks!

**Comment by lrusak** [comment]
> Yet clang format let the commit message slide? 

fixed 😉

File: xbmc/platform/darwin/osx/network/NetworkOsx.cpp [hunk 9]
```diff
@@ -59,23 +58,18 @@ bool CNetworkInterfaceOsx::GetHostMacAddress(unsigned long host_ip, std::string&
...
-  mib[5] = RTF_LLINFO;
-
-  if (sysctl(mib, ARRAY_SIZE(mib), NULL, &needed, NULL, 0) == 0)
+  if (sysctl(const_cast<int*>(mib.data()), mib.size(), NULL, &needed, NULL, 0) == 0)
   {
     char* buf = static_cast<char*>(malloc(needed));
     if (buf)
     {
-      if (sysctl(mib, ARRAY_SIZE(mib), buf, &needed, NULL, 0) == 0)
+      if (sysctl(const_cast<int*>(mib.data()), mib.size(), buf, &needed, NULL, 0) == 0)
```
Comment (AlwinEsch) [inline xbmc/platform/darwin/osx/network/NetworkOsx.cpp:72]: Maybe change here and above then direct to `nullptr`

File: xbmc/platform/freebsd/network/NetworkFreebsd.cpp (hunk 8 above)
Comment (AlwinEsch) [inline xbmc/platform/freebsd/network/NetworkFreebsd.cpp:105]: Here the same and above, as already a change inside can also come `nullptr`

File: xbmc/windowing/windows/WinKeyMap.h [hunk 10]
```diff
@@ -15,164 +15,165 @@
...
-  namespace WINDOWING
-  {
-    namespace WINDOWS
-    {
-      static XBMCKey VK_keymap[XBMCK_LAST];
+namespace WINDOWING
+{
+namespace WINDOWS
+{
+static std::array<XBMCKey, XBMCK_LAST> VK_keymap;
```
Comment (AlwinEsch) [inline xbmc/windowing/windows/WinKeyMap.h:26]: As is more in namespace, maybe a new line above.

File: xbmc/windowing/windows/WinKeyMap.h [hunk 11]
```diff
@@ -15,164 +15,165 @@
...
+    VK_keymap[VK_MEDIA_PREV_TRACK] = XBMCK_MEDIA_PREV_TRACK;
+    VK_keymap[VK_MEDIA_STOP] = XBMCK_MEDIA_STOP;
+    VK_keymap[VK_MEDIA_PLAY_PAUSE] = XBMCK_MEDIA_PLAY_PAUSE;
+    VK_keymap[VK_LAUNCH_MAIL] = XBMCK_LAUNCH_MAIL;
+    VK_keymap[VK_LAUNCH_MEDIA_SELECT] = XBMCK_LAUNCH_MEDIA_SELECT;
+    VK_keymap[VK_LAUNCH_APP1] = XBMCK_LAUNCH_APP1;
+    VK_keymap[VK_LAUNCH_APP2] = XBMCK_LAUNCH_APP2;
   }
 }
+} // namespace WINDOWS
```
Comment (AlwinEsch) [inline xbmc/windowing/windows/WinKeyMap.h:177]: Also then here a new line above.

File: xbmc/dbwrappers/mysqldataset.cpp [hunk 12]
```diff
@@ -941,7 +944,8 @@ void MysqlDatabase::mysqlVXPrintf(
     /* Fetch the info entry for the field */
     infop = &fmtinfo[0];
     xtype = etINVALID;
-    for(idx2=0; idx2<ARRAY_SIZE(fmtinfo); idx2++){
+    for (idx2 = 0; idx2 < fmtinfo.size(); idx2++)
```
Comment (AlwinEsch) [inline xbmc/dbwrappers/mysqldataset.cpp:947]: Here can also a `for (const auto& info : fmInfo)` used, maybe becomes then also `clang` more happy below (strange style there :smirk:) :smile:.

**Review by AlwinEsch — APPROVED** [review]
Only small style comments, but from my side is it OK.

File: xbmc/platform/darwin/osx/network/NetworkOsx.cpp (hunk 9 above)
Comment (lrusak) [reply xbmc/platform/darwin/osx/network/NetworkOsx.cpp]: done

File: xbmc/platform/freebsd/network/NetworkFreebsd.cpp (hunk 8 above)
Comment (lrusak) [reply xbmc/platform/freebsd/network/NetworkFreebsd.cpp]: done

File: xbmc/windowing/windows/WinKeyMap.h (hunk 10 above)
Comment (lrusak) [reply xbmc/windowing/windows/WinKeyMap.h:27]: added

File: xbmc/windowing/windows/WinKeyMap.h (hunk 11 above)
Comment (lrusak) [reply xbmc/windowing/windows/WinKeyMap.h:179]: added

File: xbmc/dbwrappers/mysqldataset.cpp (hunk 12 above)
Comment (lrusak) [reply xbmc/dbwrappers/mysqldataset.cpp]: updated with a new commit

**Review by AlwinEsch — APPROVED** [review]
Only the bad clang maybe (if it works with header on top), the rest OK.

**Pull request merged**